src/esm_runscripts/prepare.py

```diff
@@ -672,6 +674,89 @@ def set_parent_info(config):
 
+def add_vcs_info(config):
+    """
+    Adds version control system information in a plain text yaml file under the
+    experiment log directory for both the models and the esm-tools.
+
+    Parameters
+    ----------
+    config : dict
+        The experiment configuration
+
+    Returns
+    -------
+    config : dict
+        The experiment configuration
+    """
+    exp_vcs_info_file = f"{config['general']['thisrun_log_dir']}/{config['general']['expid']}_vcs_info.yaml"
+    logger.debug("Experiment information is being stored for usage under:")
+    logger.debug(f">>> {exp_vcs_info_file}")
+    vcs_versions = {}
+    all_models = config.get("general", {}).get("models", [])
+    for model in all_models:
+        model_dir = config[model]["model_dir"]
+        if helpers.is_git_repo(model_dir):
+            vcs_versions[model] = helpers.get_all_git_info(model_dir)
+        else:
+            vcs_versions[model] = "Not a git-controlled model!"
+
+    # NOTE(PG): There is no good way to get the repo directory from the config,
+    # this may at least be a good start:
+    esm_tools_repo = config.get("general", {}).get("esm_function_dir")
+    if esm_tools_repo is not None:
+        vcs_versions["esm_tools"] = helpers.get_git_hash(f"{esm_tools_repo}/../")
+    else:
+        # FIXME(PG): This should absolutely never happen. The error message could use a better wording though...
+        esm_parser.user_error("esm_tools doesn't know where it's own install location is. Something is very seriously wrong.")
+    with open(exp_vcs_info_file, "w") as f:
+        yaml.dump(vcs_versions, f)
+    return config
+
+def check_vcs_info_against_last_run(config):
+    """
+    Ensures that the version control info for two runs is identical between the
+    current run and the previous run.
+
+
+    This is checked by ensuring that the dictionary representation of the VCS
+    files is identical. Differences will result in a ``user_error``, thus
+    terminating the run. The check can be circumvented by setting::
+
+        general:
+            allow_vcs_differences: True
+
+    Parameters
+    ----------
+    config : dict
+        The experiment configuration
+
+    Returns
+    -------
+    config : dict
+        The experiment configuration
+    """
+    if config['general']['run_number'] == 1:
+        return config  # No check needed on the very first run
+    exp_vcs_info_file = f"{config['general']['thisrun_log_dir']}/{config['general']['expid']}_vcs_info.yaml"
+    # FIXME(PG): This file might not exist if people erase every run folder...
+    last_exp_vcs_info_file = f"{config['prev_run']['general']['thisrun_log_dir']}/{config['general']['expid']}_vcs_info.yaml"
+
+    with open(exp_vcs_info_file, "r") as f:
+        current_vcs_info = yaml.safe_load(f)
+    with open(last_exp_vcs_info_file, "r") as f:
+        last_vcs_info = yaml.safe_load(f)
+    if not config["general"].get("allow_vcs_differences", False) and current_vcs_info != last_vcs_info:
```

Comment:
@pgierz, I am not 100% sure but maybe library like deepdiff would be better here. It looks like `vcs_info` dicts are simple so that basic equality check should not be a big issue here.

Reply:
At the moment is a dictionary of dictionaries, nested only once. In any case, as soon as differences are detected at all, it should break, which this currently does. The only "valid" differences would be if someone goes in and changes the file by hand, without changing the actually relevant info. I currently don't think anyone would be that stupid...but, you never know. The question then becomes, how should I check for that?